bsp/src/org/jetbrains/bsp/project/test/environment/BspJvmEnvironment.scala

```diff
@@ -0,0 +1,256 @@
+import ch.epfl.scala.bsp4j.JvmRunEnvironmentParams
+import ch.epfl.scala.bsp4j.JvmRunEnvironmentResult
+import ch.epfl.scala.bsp4j.JvmTestEnvironmentParams
+import ch.epfl.scala.bsp4j.JvmTestEnvironmentResult
+import ch.epfl.scala.bsp4j.SourceItem
+import ch.epfl.scala.bsp4j.SourcesItem
+import ch.epfl.scala.bsp4j.SourcesParams
+import ch.epfl.scala.bsp4j.SourcesResult
+import com.intellij.execution.configurations.ModuleBasedConfiguration
+import com.intellij.execution.configurations.RunConfiguration
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.externalSystem.util.ExternalSystemApiUtil
+import com.intellij.openapi.module.Module
+import com.intellij.openapi.project.Project
+import com.intellij.psi.JavaPsiFacade
+import com.intellij.psi.PsiFile
+import com.intellij.psi.search.GlobalSearchScope
+import org.jetbrains.bsp.BspBundle
+import org.jetbrains.bsp.BspUtil._
+import org.jetbrains.bsp.data.BspMetadata
+import org.jetbrains.bsp.protocol.BspCommunication
+import org.jetbrains.bsp.protocol.BspJob
+import org.jetbrains.bsp.protocol.session.BspSession.BspServer
+import org.jetbrains.bsp.protocol.session.BspSession.BspSessionTask
+import org.jetbrains.plugins.scala.build.BuildMessages
+import org.jetbrains.plugins.scala.build.BuildToolWindowReporter
+import org.jetbrains.plugins.scala.build.BuildToolWindowReporter.CancelBuildAction
+import org.jetbrains.plugins.scala.extensions.inReadAction
+import org.jetbrains.plugins.scala.extensions.invokeAndWait
+
+import scala.collection.JavaConverters._
+import scala.concurrent.Promise
+import scala.util.Failure
+import scala.util.Success
+import scala.util.Try
+
+object BspJvmEnvironment {
+
+  trait BspTargetIdHolder {
+    def currentValue: Option[BuildTargetIdentifier]
+    def update(value: BuildTargetIdentifier)
+  }
+
+  case class Error(msg: String) extends RuntimeException(msg)
+
+  type Result[A] = Either[Error, A]
+
+  def getBspTargets(module: Module): Result[Seq[BuildTargetIdentifier]] = {
+    BspMetadata.get(module.getProject, module)
+      .left.map(err => Error(err.msg))
+      .map(data => data.targetIds.asScala.map(id => new BuildTargetIdentifier(id.toString)))
+  }
+
+  def promptUserToSelectBspTarget(
+    project: Project,
+    targetIds: Seq[BuildTargetIdentifier],
+    holder: BspTargetIdHolder
+  ): Option[BuildTargetIdentifier] = {
+    val selected = invokeAndWait(BspSelectTargetDialog.promptForBspTarget(project, targetIds, holder.currentValue))
+    selected.foreach(holder.update)
+    selected
+  }
+
+  def promptUserToSelectBspTargetForWorksheet(module: Module): Unit = {
+    val potentialTargets = getBspTargets(module)
+    potentialTargets.foreach { targetIds =>
+      val holder = persistentHolderForWorksheet(module)
+      promptUserToSelectBspTarget(module.getProject, targetIds, holder)
+    }
+  }
+
+  def resolveForWorksheet(module: Module): Result[JvmEnvironment] = {
+    val holder = persistentHolderForWorksheet(module)
```

Comment:
Does it mean that it is impossible to have two worksheets of the same module, but using different targets?

Reply:
Yeah, if someone has 2 worksheets that use module that has multiple targets in it, they will share the same target. If someone was switching between those, would need to change the target before run. But I hope that for now it's not a common use case.